### ISSUE TITLE
backtest opens with trade open (open open) bug

### ISSUE THREAD

**Issue by creslinux** [freqtrade/freqtrade#860]
This issue was found during testing of PR 855 but is a standalone issue (probably) so im raising a separate ticket.  #855 

Freqtrade backtest has a bug
No matter the timerange passed freqtrade opens with a trade in an open state. 

I have applied the lookahead patch from #855 

For timeranges yyyymmdd-yyyymmdd these false trades are open at midnight exactly. 
I've also passed a timestamp range from 18:00 and it also starts with a trade open. 

Subsequent trades found by FreqTrade in the range align perfectly with what Tradingview lists. 
The start with a trade open bug throws the profit/drawdown results for the period. 

Examples of the **(open open)** bug and aligning with TV. 
<img width="1036" alt="screen shot 2018-06-06 at 8 16 24 pm" src="https://user-images.githubusercontent.com/34645187/41063854-691af5dc-69c9-11e8-980d-c36850b1e898.png">
<img width="737" alt="screen shot 2018-06-06 at 8 12 12 pm" src="https://user-images.githubusercontent.com/34645187/41063855-6940406c-69c9-11e8-8f82-68aaf0fa2f53.png">

<img width="771" alt="screen shot 2018-06-06 at 8 12 49 pm" src="https://user-images.githubusercontent.com/34645187/41063864-6ff429c8-69c9-11e8-955a-cb9206be3667.png">
<img width="737" alt="screen shot 2018-06-06 at 8 13 15 pm" src="https://user-images.githubusercontent.com/34645187/41063865-701742dc-69c9-11e8-8269-ed4d977f402b.png">

Example using timestamp to set the range to 18:00 start. 
<img width="824" alt="screen shot 2018-06-06 at 8 27 18 pm" src="https://user-images.githubusercontent.com/34645187/41063900-83a8a534-69c9-11e8-9fe8-79edbf6fcd59.png">

**Comment by creslinux** [comment]
The strategy used to produce this condition is included below. 
This was ran against binance on 1hr candles for ETH/BTC

```
# Freqtrade_backtest_validation_freqtrade1.py
# This script is 1 of a pair the other being freqtrade_backtest_validation_tradingview1
# These should be executed on their respective platforms for the same coin/period/resolution 
# The purpose is to test Freqtrade backtest provides like results to a known industry platform. 
#
# --- Do not remove these libs ---
from freqtrade.strategy.interface import IStrategy
from pandas import DataFrame
# --------------------------------

# Add your lib to import here
import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib
import numpy 

class Freqtrade_backtest_validation_freqtrade1(IStrategy):
    # Minimal ROI designed for the strategy.
    # This attribute will be overridden if the config file contains "minimal_roi"
    minimal_roi = {
        "0": 0.5
    }

    # Optimal stoploss designed for the strategy
    # This attribute will be overridden if the config file contains "stoploss"
    stoploss = -0.2

    # Optimal ticker interval for the strategy
    ticker_interval = '4h'

    def populate_indicators(self, dataframe: DataFrame) -> DataFrame:

        dataframe['smaFast'] = ta.SMA(dataframe, timeperiod=14)
        dataframe['smaSlow'] = ta.SMA(dataframe, timeperiod=28)

        return dataframe

    def populate_buy_trend(self, dataframe: DataFrame) -> DataFrame:
        """
        Based on TA indicators, populates the buy signal for the given dataframe
        :param dataframe: DataFrame
        :return: DataFrame with buy column
        """
        dataframe.loc[
            (
                qtpylib.crossed_above(dataframe['smaFast'], dataframe['smaSlow'])
            ),
            'buy'] = 1

        return dataframe

    def populate_sell_trend(self, dataframe: DataFrame) -> DataFrame:
        """
        Based on TA indicators, populates the sell signal for the given dataframe
        :param dataframe: DataFrame
        :return: DataFrame with buy column
        """
        dataframe.loc[
            (
                qtpylib.crossed_below(dataframe['smaFast'], dataframe['smaSlow'])
            ),
            'sell'] = 1
        return dataframe
```

**Comment by vertti** [comment]
Excellent bug report. Very very important research.

**Comment by xmatthias** [comment]
@creslinux i think the you're experiencing is that if you use 20180520-20180527 it only loads data for exactly this period. So it won't have valid indicators (depending which ones you use) for the first few hours.

we should load another X candles before "trade start" ... which is how the bot "really" behaves  - if i start it at midnight exactly, it will download 100 (or so) candles from the previous day - which allows it to start analyzing immediately.

so if you specify 20180620- - it should actually load data from 20180619_18:00:00 but only generates buy-signals after midnight - currently i think it won't generate buy-signals until 2018-06-20_06:00:00 due to not having history for sma indicators.

(please note that i did not fully investigate this yet - the times are guesses only, could also only be 4 hours  - depends a lot on the indicators and timeframe used).

**Comment by vertti** [comment]
You will need atleast as many candles than the longest window of your SMA's is before they are accurate.

**Comment by creslinux** [comment]
Is it as simple as cancel, or better phrased, do not lodge/honour/ any order made the same second bot starts?

**Comment by xmatthias** [comment]
i'm not sure i understand what you mean ... in the last 2 comments.
what we're saying is that if you backtest from 20180601 to 20180605 with 1h candles and the highest sma timeframe is 28:
you will need `28 * 1h = 1.2 days` of "startuptime"-data to actually have SMA populated correctly. so if you want the first trade to happen ad midnight 20180601 - you need to have data from 20180530_20:00  onwards - otherwise SMA (28) is empty for 28 ticks (28hours in this case).

of course - if you use SMA90, you will need 90 hours of "startup-data".

ta-lib simply outputs NAN until the timeframe is available - unless you specifically check for that - no trade should happen until the startup-time is complete.

**Comment by creslinux** [comment]
The 200day MA and 55MA are commonly used. 
A golden / death cross being when they crossover-under that may occur once a year
a lot of markets wont have that depth of history as they were introduced to the exchange post a year 

If we know there is a 100% hit that any trade opened on the first tick/second of a test/trade period is false - isn't the correct means to manage that to not process any trades on opening to backtest or live if the open timestamp is the same as backtest period start or bot start

**Comment by xmatthias** [comment]
i think you misunderstand what i mean
it's another difference between backtest and real/dry trading mode.

real/dry gets the last X candles right at startup and analyzes that - so has the indicator available for the current time.

backtest however, you define a first tick (midnight of a day - in our case) **loaded** - so the first buy-signal will be generated after first tick + largest indicator offset used. (potentially at 6am in the morning - or the day after. It depends on timeframe + indicator offset used.

**Comment by creslinux** [comment]
If I have a 55/200 daily I need 200 days of data.
If the coin launched yesterday on the exchange I think we hit the same as backtesting here

Coins launch each few days on binance so lots of shorter term strategies could fall foul.

**Comment by creslinux** [comment]
@xmatthias  as you say live appears protected. (i tested, not that i don't trust you)

I ran a dry-run against a coin listed on binance (NXT/BTC) yesterday; for an 1hr SMA 56 /112 crossover. 
No open trade was launched on bot start even though the ticker history did not precede the SMA test 

> (.env) DannyMBP:freqtrade_new creslin$  python3 ./freqtrade/main.py -v --strategy Freqtrade_live_test
> 2018-06-07 10:10:46,027 - freqtrade.configuration - INFO - Log level set to DEBUG
> 2018-06-07 10:10:46,028 - freqtrade.configuration - DEBUG - Exchange "binance" supported
> 2018-06-07 10:10:46,028 - freqtrade.configuration - INFO - Using max_open_trades: 200 ...
> 2018-06-07 10:10:46,029 - freqtrade.configuration - INFO - Using data folder: user_data/data/binance ...
> 2018-06-07 10:10:46,029 - freqtrade.freqtradebot - INFO - Starting freqtrade 0.17.0
> 2018-06-07 10:10:46,029 - freqtrade.strategy.resolver - DEBUG - Searching for strategy Freqtrade_live_test in '/Users/creslin/PycharmProjects/freqtrade_new/freqtrade/strategy/../../user_data/strategies'
> 2018-06-07 10:10:46,035 - freqtrade.strategy.resolver - INFO - Using resolved strategy Freqtrade_live_test from '/Users/creslin/PycharmProjects/freqtrade_new/freqtrade/strategy/../../user_data/strategies'
> 2018-06-07 10:10:46,035 - freqtrade.strategy.resolver - INFO - Override strategy 'ticker_interval' with value in config file: 1h.
> 2018-06-07 10:10:46,199 - freqtrade.exchange - INFO - Instance is running with dry_run enabled
> 2018-06-07 10:10:46,202 - freqtrade.exchange - INFO - Using Exchange "Binance"
> 2018-06-07 10:10:47,216 - freqtrade.rpc.rpc_manager - INFO - *Status:* `running`
> 2018-06-07 10:10:47,216 - freqtrade.freqtradebot - INFO - Changing state to: RUNNING
> 2018-06-07 10:10:48,152 - freqtrade.freqtradebot - INFO - Checking buy signals to create a new trade with stake_amount: 1.000000 ...
> 2018-06-07 10:10:48,451 - freqtrade.exchange - DEBUG - Downloaded data for NXS/BTC time range [2018-06-06 10:00:00+00:00, 2018-06-07 10:00:00+00:00]
> 2018-06-07 10:10:48,491 - freqtrade.analyze - DEBUG - trigger: 2018-06-07 10:00:00+00:00 (pair=NXS/BTC) buy=False sell=False
> 2018-06-07 10:10:48,492 - freqtrade.freqtradebot - INFO - Found no buy signals for whitelisted currencies. Trying again..
> 2018-06-07 10:10:48,495 - freqtrade.freqtradebot - DEBUG - Throttling _process for 3.72 seconds
> 2018-06-07 10:10:52,527 - freqtrade.freqtradebot - INFO - Checking buy signals to create a new trade with stake_amount: 1.000000 ...
> 2018-06-07 10:10:53,020 - freqtrade.exchange - DEBUG - Downloaded data for NXS/BTC time range [2018-06-06 10:00:00+00:00, 2018-06-07 10:00:00+00:00]
> 2018-06-07 10:10:53,036 - freqtrade.analyze - DEBUG - trigger: 2018-06-07 10:00:00+00:00 (pair=NXS/BTC) buy=False sell=False
> 2018-06-07 10:10:53,036 - freqtrade.freqtradebot - INFO - Found no buy signals for whitelisted currencies. Trying again..
>

**Comment by xmatthias** [comment]
testing this should be fairly easy - add a print in the strategy for the first item in the dataframe - and the first which doesn't have the indicator NAN and compare the timestamp ... ;)
this can be compared between live and backtest ... and "should" show that.

my work-network doesn't allow me to run freqtrade - so i can't test that now  - but in theory it should work ...

**Comment by vertti** [comment]
@xmatthias i think that check is valid for MA's.. but several other indicators have what TA-lib calls unstable period.. so it will have values but they are not the "correct" values for a while before it has enough history data.

**Comment by creslinux** [comment]
I think #855 closed this?
